performjoin.go

```diff
@@ -0,0 +1,276 @@
+package gomatrixserverlib
+
+import (
+	"context"
+	"crypto/ed25519"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/matrix-org/gomatrixserverlib/spec"
+	"github.com/sirupsen/logrus"
+)
+
+type PerformJoinInput struct {
+	UserID     *spec.UserID
+	RoomID     string
+	ServerName spec.ServerName
+	Content    map[string]interface{}
+	Unsigned   map[string]interface{}
+
+	PrivateKey ed25519.PrivateKey
+	KeyID      KeyID
+	KeyRing    *KeyRing
+
+	EventProvider EventProvider
+}
+
+type PerformJoinResponse struct {
+	JoinEvent     *HeaderedEvent
+	StateSnapshot StateResponse
+}
+
+// PerformJoin provides high level functionality that will attempt a federated room
+// join. On success it will return the new join event and the state snapshot returned
+// as part of the join.
+func PerformJoin(
+	ctx context.Context,
+	fedClient FederatedJoinClient,
+	input PerformJoinInput,
+) (*PerformJoinResponse, *FederationError) {
+	origin := input.UserID.Domain()
+
+	// Try to perform a make_join using the information supplied in the
+	// request.
+	respMakeJoin, err := fedClient.MakeJoin(
+		ctx,
+		origin,
+		input.ServerName,
+		input.RoomID,
+		input.UserID.Raw(),
+	)
+	if err != nil {
+		// TODO: Check if the user was not allowed to join the room.
+		return nil, &FederationError{
+			ServerName: input.ServerName,
+			Transient:  true,
+			Reachable:  false,
```

Comment:
This assumes a network error, but I'm fairly sure we will return an `err` on non-2xx as well.

Reply:
Yup. This keeps the behaviour the same as before the refactor. I tried to limit behaviour changes in this PR